nomad/encrypter.go

```diff
@@ -341,21 +344,147 @@ func (e *Encrypter) VerifyClaim(tokenString string) (*structs.IdentityClaims, er
+		// stop any previous tasks for this same key ID under the assumption
+		// they're broken or being superceded, but don't remove the CancelFunc
+		// from the map yet so that other callers don't think we can continue
+		cancel()
+	}
+
+	e.lock.Unlock()
+
+	completeCtx, cancel := context.WithCancel(ctx)
+
+	for _, wrappedKey := range wrappedKeys.WrappedKeys {
+		providerID := wrappedKey.ProviderID
+		if providerID == "" {
+			providerID = string(structs.KEKProviderAEAD)
+		}
+
+		provider, ok := e.providerConfigs[providerID]
+		if !ok {
+			logger.Error("no such KMS provider configured - root key cannot be decrypted",
+				"provider_id", providerID)
+			cancel()
+			return fmt.Errorf("no such provider %q configured", providerID)
+		}
+
+		wrapper, err := e.newKMSWrapper(provider, wrappedKeys.KeyID, wrappedKey.KeyEncryptionKey)
+		if err != nil {
+			// the errors that bubble up from this library can be a bit opaque, so
+			// make sure we wrap them with as much context as possible
+			logger.Error("unable to create KMS wrapper - root key cannot be decrypted",
+				"provider_id", providerID, "error", err)
+
+			cancel()
+			return fmt.Errorf("unable to create key wrapper for provider %q: %w", providerID, err)
+		}
+
+		// fan-out decryption tasks for HA in Nomad Enterprise. we can use the
+		// key whenever any one provider returns a successful decryption
+		go e.decryptWrappedKeyTask(completeCtx, cancel, wrapper, provider, wrappedKeys.Meta(), wrappedKey)
 	}
+
+	e.lock.Lock()
+	defer e.lock.Unlock()
+
+	e.decryptTasks[wrappedKeys.KeyID] = cancel
+
 	return nil
 }
 
+func (e *Encrypter) decryptWrappedKeyTask(ctx context.Context, cancel context.CancelFunc, wrapper kms.Wrapper, provider *structs.KEKProviderConfig, meta *structs.RootKeyMeta, wrappedKey *structs.WrappedRootKey) {
+
+	var key []byte
+	var rsaKey []byte
+	var err error
+
+	minBackoff := time.Second
+	maxBackoff := time.Second * 5
+
+	helper.WithBackoffFunc(ctx, minBackoff, maxBackoff, func() error {
+		wrappedDEK := wrappedKey.WrappedDataEncryptionKey
+		key, err = wrapper.Decrypt(e.srv.shutdownCtx, wrappedDEK)
+		if err != nil {
+			err := fmt.Errorf("%w (root key): %w", ErrDecryptFailed, err)
+			e.log.Error(err.Error(), "key_id", meta.KeyID)
+			return err
+		}
+		return nil
+	})
+
+	helper.WithBackoffFunc(ctx, minBackoff, maxBackoff, func() error {
+		// Decrypt RSAKey for Workload Identity JWT signing if one exists. Prior to
+		// 1.7 an ed25519 key derived from the root key was used instead of an RSA
+		// key.
+		if wrappedKey.WrappedRSAKey != nil {
+			rsaKey, err = wrapper.Decrypt(e.srv.shutdownCtx, wrappedKey.WrappedRSAKey)
+			if err != nil {
+				err := fmt.Errorf("%w (rsa key): %w", ErrDecryptFailed, err)
+				e.log.Error(err.Error(), "key_id", meta.KeyID)
+			}
+		}
+		return nil
+	})
+
+	rootKey := &structs.RootKey{
+		Meta:   meta,
+		Key:    key,
+		RSAKey: rsaKey,
+	}
+
+	helper.WithBackoffFunc(ctx, minBackoff, maxBackoff, func() error {
+		err = e.addCipher(rootKey)
```

Comment:
`key` and/or `rsaKey` may still be `nil` here since we never exit early even if Decrypt fails.

Reply:
We need to early return on an error from any of these WithBackoffFuncs I think. Otherwise we risk adding invalid keys and/or treating failures as successes.

Reply:
My thinking here was that the context is always the server's shutdown context but (a) that's brittle to change, and (b) just because the server shutdown context has been called doesn't mean that functions downstream of this won't get called with whatever side-effects they have. Will fix.